22 UI bug

Julius-Babies opened the pull request:
I will review the changes now

Belgarus committed:
added logs for toubleshooting

Belgarus committed:
fixed first few issues

Belgarus committed:
fixed issues but its still loading

Belgarus committed:
replaced the threads in favor of Kotlin Coroutines

Belgarus committed:
current status, I dont know how to fix it at the moment

Belgarus committed:
fixed the UI-Problem

Nnamllit1 committed:
Create updateStops

Belgarus committed:
major changes

Belgarus committed:
Merge remote-tracking branch 'origin/22-ui-bug' into 22-ui-bug

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/MainActivity.kt:
```diff
@@ -190,136 +138,238 @@ class MainActivity : AppCompatActivity() {
                 provider,
-                1000L,  // 1 Sekunde
-                1f,     // 1 Meter
+                1000L, 
```
So-called "Magic-numbers" should be defined using a constant at a somewhat top level given the importance of those. See https://en.wikipedia.org/wiki/Magic_number_(programming)?useskin=vector

Julius-Babies replied:
Also include the unit in the constant name. I suggest using Kotlin's `const`-keyword (see https://kotlinlang.org/docs/properties.html#compile-time-constants)

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/MainActivity.kt:
This file seems rather large, consider refactoring some components into useful separate files enhancing reusability and testability in the future. However, this is not necessary yet so feel free to just skip this :)

Julius-Babies reviewed (commented):
Small changes, someone else needs to approve this since I started this pr and therefore cannot approve it.

Belgarus committed:
fix: improve TTS announcements and station detection - Increase station detection radius from 200m to 250m for better user experience - Fix S-Bahn line announcements to avoid Gleis prefix - Improve platform name handling to remove redundant Gleis prefixes - Fix German text for eine Minute delay announcement - Refactor line announcement logic for different transport types: S-Bahn: S1 in Richtung X an Gleis Y, Trains: Gleis RE1 in Richtung X an Gleis Y, Bus/Tram: Linie 62 in Richtung X an Steig Y

Belgarus commented:
Okay, I’ll keep working on this a bit more, but right now everything seems to be working fine.
<img width="275px" src="https://github.com/user-attachments/assets/d0bc7f02-7e74-4bb5-be00-023e1bf0e104">

Belgarus committed:
the train-delay is now displayed on the UI

Belgarus committed:
refactor: replace magic numbers with named constants - Add companion object with location-related constants - Add provider-specific time and distance thresholds - Add sensor-related constants - Improve code maintainability and readability - Make configuration values easier to modify

Belgarus committed:
refactor: improve constant naming and use const keyword - Add unit suffixes to constant names (METERS, MILLIS, NANOS) - Use Kotlin const keyword for compile-time constants - Remove redundant comments as units are now in names - Update all constant references throughout the code - Improve code readability and maintainability

Belgarus committed:
refactor: Improve UI and station display

Refactor MainActivity to use Compose and MVVM architecture. Adjust Play/Pause button to take up 50% of screen height and optimize station departure display to show only next departure per line. Add proper delay indicators and improve loading states. Fix location permission handling and station distance calculation.

Belgarus commented:
# My main changes:
```

- Refactor MainActivity to use Compose and MVVM architecture
- Adjust Play/Pause button to 50% screen height
- Show only next departure per line/destination
- Add proper delay indicators (+/-) for departures
- Fix location permission handling
- Fix station distance calculation

```

https://github.com/user-attachments/assets/3d1d21a5-4459-4b62-8edd-e1e7738c452a

TTS works perfectly fine, but sound doesn’t seem to work in the markup, I guess.

Belgarus commented:
Should I simplify the files a bit more? It’s quite a lot of code.

Belgarus commented:
Would you mind testing if the app works on your device too? I’ve only tried it on my old Android phone so far.

Julius-Babies commented:
> Should I simplify the files a bit more? It’s quite a lot of code.

Next time, split these changes into multiple tickets and pull requests (PRs). I saw something related to translations in the changes that has nothing to do with the initial issue #22. It's really hard to review these massive PRs, that's why it takes so long.

* * *

> Would you mind testing if the app works on your device too? I’ve only tried it on my old Android phone so far.

If it works on your device, it should also work on newer ones. If you're able to test it on an emulator with a more recent version of Android, feel free to do so. If the app works on there too, I'll merge this and we forget that this big PR ever existed ;)

Belgarus commented:
Okay I will test it and will look that everything works fine. :)

Belgarus committed:
fix: splash screen drawable crash

Belgarus commented:
The Emulator is stuck loading. Should I open a new PR? 

<img width="300px" src="https://github.com/user-attachments/assets/35b61eef-bfcc-4aff-81b9-3755d81f73b1">

Belgarus committed:
fixed DefaultLocale warning

Belgarus commented:
Can we merge this as an unstable version?

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/dvb/DVB.kt:
```diff
@@ -75,7 +80,7 @@ class DVBSource(
                         val plannedTime = extractLocalTimeFromDateString(it.scheduleTime)
                         val actualTime = extractLocalTimeFromDateString(realTime)
-                        (plannedTime.hour - actualTime.hour) * 60 + (plannedTime.minute - actualTime.minute)
+                        (actualTime.hour - plannedTime.hour) * 60 + (actualTime.minute - plannedTime.minute)
```
Maybe we should switch to kotlinx.datetime to get access to the Duration class which allows something like `(actualTime - plannedTime).inWholeMinutes`.

If you want to work on something, you should check the library out: https://github.com/Kotlin/kotlinx-datetime

This would be a required step to make the app work on iOS, since the LocalDate class we are currently using is a Java-Class which we are unable to utilize when targeting non-Android platforms. Feel free to create a master ticket for Kotlin Multiplatform for iOS (see https://kotlinlang.org/docs/multiplatform/kmp-overview.html) and a child ticket for the replacement of java DateTime API with the Kotlin alternative.

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/dvb/DVB.kt:
```diff
@@ -86,16 +91,59 @@ class DVBSource(
+                        emptyList<Station>()
+                    } else {
+                        Log.d("DVBSource", "JSON content length: ${content.length}")
+                        Log.d("DVBSource", "First 100 characters: ${content.take(100)}")
```
We don't need to log this

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/dvb/DVB.kt:
```diff
@@ -86,16 +91,59 @@ class DVBSource(
-                    Station(it.properties.number, it.properties.name, it.geometry.coordinates[0], it.geometry.coordinates[1], emptyList())
+        Log.d("DVBSource", "Starting to load stations from raw resource")
+        return try {
+            val inputStream = context.resources.openRawResource(R.raw.stops)
```
This should be deferred to the VVO-Api in the long term, so we can also get rid of the batch script which generates the json. Feel free to create a ticket.

Julius-Babies commented on app/src/main/java/org/jugendhackt/wegweiser/SplashActivity.kt:
Do we need a dedicated activity for this? We have a loading indicator on the main activity

Nnamllit1 replied:
I have no idea why the splash screen exists, because it only delays the app startup by 1.5s :)

Julius-Babies commented on updateStops:
Either complete or remove this file

Julius-Babies reviewed (commented):
Since I opened this PR, I cannot approve nor request changes. If you clean up the unnecessary stuff we can merge this since we have to do some architectural planning anyways later on.

Julius-Babies committed:
Merge branch 'main' into 22-ui-bug

Commit:
Implement automated station data refreshing and caching

Replaced the manual Python-based stop update process with a `StationStore` that handles periodic network fetching and local caching. Updated the `MainViewModel` to observe dataset changes and improved location update handling with job cancellation.

- Added `StationStore` to manage stop data, local caching, and 24-hour refresh logic.
- Removed legacy `main.py` and batch update scripts.
- Updated `DVBSource` to delegate station management to `StationStore`.
- Integrated automatic station refreshing into `MainApplication` startup.
- Refactored `MainViewModel` to react to station dataset updates and handle location updates more efficiently.

Commit:
Add comments explaining StationStore loading and caching logic

Commit:
Inject HttpClient and simplify StationStore

- Refactored `DVBSource` and `StationStore` to receive `HttpClient` via dependency injection.
- Added `HttpClient` with the `CIO` engine to the Koin `appModule`.
- Removed `AtomicReference` and synchronous `getStations()` in favor of `StateFlow`.
- Updated `MainViewModel` to retrieve stations using `observeStations().first()`.
- Simplified `StationStore` by converting helper methods to properties and renaming `shouldRefresh` to `isCacheExpired`.

Commit:
Document drop(1) usage in station observer

Added a comment explaining that the initial emission is skipped to only react to later dataset updates.

Commit:
Automate stop data generation

Added a Python script to fetch and convert VVO stop data to GeoJSON and integrated its execution into the CI build workflow. Refreshed stop names, locations, and coordinates in `stops.json`.

Nnamllit1 committed:
Merge pull request #29 from Julius-Babies/28-updating-stops-in-runtime

28-updating-stops-in-runtime

Nnamllit1 commented:
@Julius-Babies What do we do with this branch?

Commit:
Migrate time handling to kotlinx-datetime

Replaced java.time usage with kotlinx-datetime for better Kotlin compatibility. This includes:
* Adding the kotlinx-datetime dependency.
* Updating delay and departure time calculations in DVB.kt and MainViewModel.kt using Kotlin's Duration API.
* Implementing a helper for retrieving the current local time via Clock.System.

Nnamllit1 commented on app/src/main/java/org/jugendhackt/wegweiser/MainViewModel.kt:
```diff
@@ -108,46 +206,96 @@ data class Station(
+                        append("${language.getString("tts.line")} ${departure.line}")
+                    }
                 }
-                append(it.line.replace(Regex("(?<=[A-Za-z])(?=\\d)"), " ")) // to make sure that the line is spoken correctly ("S80" -> "S 80")
```
Why got this changed?

Julius-Babies replied:
I think I split that up because "S80" would be spelled as "S acht null" instead of "S Achtzig"

Nnamllit1 replied:
ok :)

Julius-Babies commented:
> @Julius-Babies What do we do with this branch?

Since it is fairly big and has conflicts with `main` I think abandoning it would be a good idea but we should pick a few changes to their own issues and PRs. [The](https://github.com/Julius-Babies/JH_DVBerry/pull/25#discussion_r3173765309) is an example for that - that could me moved to its own issue.

Nnamllit1 commented:
>Since it is fairly big and has conflicts with `main` I think abandoning it would be a good idea but we should pick a few changes to their own issues and PRs. [The](https://github.com/Julius-Babies/JH_DVBerry/pull/25#discussion_r3173765309) is an example for that - that could me moved to its own issue. 

Ok, woud you pick and move the changes?

Julius-Babies commented:
Sure, I can take a look into it this afternoon
